lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3639,25 +3630,89 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
 
+// Return the expected generic signature for autodiff associated functions given
+// a SILDifferentiableAttr. The expected generic signature is built from the
+// original generic signature and the attribute's requirements.
+static GenericSignature *
+getAutoDiffAssociatedFunctionGenericSignature(SILDifferentiableAttr *attr,
+                                              SILFunction *original) {
+  auto originalGenSig =
+      original->getLoweredFunctionType()->getGenericSignature();
+  if (!originalGenSig)
+    return nullptr;
+  GenericSignatureBuilder builder(original->getASTContext());
+  // Add original generic signature.
+  builder.addGenericSignature(originalGenSig);
+  // Add where clause requirements.
+  auto source =
+      GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+  for (auto &req : attr->getRequirements())
+    builder.addRequirement(req, source, original->getModule().getSwiftModule());
+  auto canGenericSig = std::move(builder)
+                           .computeGenericSignature(
+                               SourceLoc(), /*allowConcreteGenericParams=*/true)
+                           ->getCanonicalSignature();
+  return canGenericSig;
+}
+
 DifferentiationTask::DifferentiationTask(ADContext &context,
                                          SILFunction *original,
                                          SILDifferentiableAttr *&&attr,
                                          DifferentiationInvoker invoker)
     : context(context), original(original), attr(attr), invoker(invoker) {
+  auto &module = context.getModule();
+  auto originalTy = original->getLoweredFunctionType();
+  auto originalLoc = original->getLocation();
   if (attr->hasJVP()) {
-    jvp = lookUpOrLinkFunction(attr->getJVPName(), context.getModule());
-    assert(jvp);
+    // If attribute specifies JVP name, try to look up JVP in current module.
+    // Otherwise, create an external reference.
+    jvp = module.lookUpFunction(attr->getJVPName());
+    if (!jvp) {
+      auto jvpName = attr->getJVPName();
+      auto *jvpGenericSig = getAutoDiffAssociatedFunctionGenericSignature(attr, original);
+      auto *jvpGenericEnv = jvpGenericSig
+          ? jvpGenericSig->createGenericEnvironment()
+          : nullptr;
+      auto jvpType = originalTy->getAutoDiffAssociatedFunctionType(
+          getIndices().parameters, getIndices().source, 1,
+          AutoDiffAssociatedFunctionKind::JVP, module,
+          LookUpConformanceInModule(module.getSwiftModule()), jvpGenericSig);
+      SILOptFunctionBuilder fb(context.getTransform());
+      jvp = fb.createFunction(SILLinkage::PublicExternal, jvpName, jvpType,
+                              jvpGenericEnv, originalLoc, original->isBare(),
+                              IsNotTransparent, original->isSerialized());
+      // NOTE: Setting debug scope is necessary to prevent crash in TFPartition.
+      jvp->setDebugScope(new (module) SILDebugScope(originalLoc, jvp));
```

Comment:
To elaborate: [`TensorFunctionClassifier::shouldBePartitioned(SILFunction *fn, bool forceTFFunctions)`](https://github.com/apple/swift/blob/b1a9dae019604daa9525a6e6a4718a643178cdee/lib/SILOptimizer/Mandatory/TFUtilities.cpp#L401) calls `fn->getDeclContext()` without first checking whether `fn` has a valid debug scope, causing a crash.

However, `SILFunction`s should have debug scopes, so it's reasonable to assume they do have one.